src/polkadot/fellows.ts

```diff
@@ -0,0 +1,109 @@
+    api = await ApiPromise.create({ provider: wsProvider });
+    try {
+      // We fetch all the members
+      const membersObj = await api.query.fellowshipCollective.members.entries();
+
+      // We iterate over the fellow data and convert them into usable values
+      const fellows: FellowData[] = [];
+      for (const [key, rank] of membersObj) {
+        // @ts-ignore
+        const [address]: [string] = key.toHuman();
+        fellows.push({ address, ...(rank.toHuman() as object) } as FellowData);
+      }
+      this.logger.info(JSON.stringify(fellows));
+
+      // Once we obtained this information, we disconnect this api.
+      await api.disconnect();
+
+      this.logger.debug("Connecting to relay parachain.");
+      // We connect to the relay chain
+      api = await ApiPromise.create({ provider: new WsProvider("wss://rpc.polkadot.io") });
+
+      // We iterate over the different members and extract their data
+      const users: Map<string, number> = new Map<string, number>();
+      for (const fellow of fellows) {
+        this.logger.debug(`Fetching identity of '${fellow.address}', rank: ${fellow.rank}`);
+        const fellowData = (await api.query.identity.identityOf(fellow.address)).toHuman() as
+          | Record<string, unknown>
+          | undefined;
+
+        // If the identity is null, we ignore it.
+        if (!fellowData) {
+          this.logger.debug("Identity is null. Skipping");
+          continue;
+        }
+
+        // @ts-ignore
+        const additional = fellowData.info.additional as [{ Raw: string }, { Raw: string }][] | undefined;
+
+        // If it does not have additional data (GitHub handle goes here) we ignore it
+        if (!additional || additional.length < 1) {
+          this.logger.debug("Additional data is null. Skipping");
+          continue;
+        }
+
+        for (const additionalData of additional) {
+          const [key, value] = additionalData;
+          // We verify that they have an additional data of the key "github"
+          // If it has a handle defined, we push it into the array
+          if (key?.Raw && key?.Raw === "github" && value?.Raw && value?.Raw.length > 0) {
+            this.logger.debug(`Found handles: '${value.Raw}'`);
+            // We add it to the array and remove the @ if they add it to the handle
+            users.set(value.Raw.replace("@", ""), fellow.rank);
+          }
+        }
+      }
+
+      this.logger.info(`Found users: ${JSON.stringify(Array.from(users.entries()))}`);
+
+      // We disconnect the API before returning the object
+      await api.disconnect();
+      return users;
+    } catch (error) {
+      console.error(error);
+      await api.disconnect();
+      throw error;
+    }
+  }
+
+  async getTeamMembers(ranking: string): Promise<string[]> {
+    const requiredRank = Number(ranking);
+    this.logger.info(`Fetching members of rank '${requiredRank}' or higher`);
+
+    if (this.fellowsCache.size < 1) {
```

Comment:
this will use data from cache in case if there are more than 1 rule bucket with rank: x?
i don't see any other use case then... right?

Reply:
It is a map. The idea for this is to fetch **all** the available fellows once and then we filter per request, instead of fetching them multiple times per required rank.

So if you request in one rule a rank 4, it will already fetch them all, so, if you request later rank 2, all the ranks will be available already.